src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/StringMarshallingInfoProvider.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Interop
+{
+    public sealed class StringMarshallingInfoProvider : ITypeBasedMarshallingInfoProvider
+    {
+        private readonly Compilation _compilation;
+        private readonly IGeneratorDiagnostics _diagnostics;
+        private readonly AttributeData _stringMarshallingCustomAttribute;
+        private readonly DefaultMarshallingInfo _defaultMarshallingInfo;
+
+        public StringMarshallingInfoProvider(Compilation compilation, IGeneratorDiagnostics diagnostics, AttributeData stringMarshallingCustomAttribute, DefaultMarshallingInfo defaultMarshallingInfo)
+        {
+            _compilation = compilation;
+            _diagnostics = diagnostics;
+            _stringMarshallingCustomAttribute = stringMarshallingCustomAttribute;
+            _defaultMarshallingInfo = defaultMarshallingInfo;
+        }
+
+        public bool CanProvideMarshallingInfoForType(ITypeSymbol type) => type.SpecialType == SpecialType.System_String;
+
+        public MarshallingInfo GetMarshallingInfo(ITypeSymbol type, int indirectionDepth, UseSiteAttributeProvider useSiteAttributes, GetMarshallingInfoCallback marshallingInfoCallback)
+        {
+            if (_defaultMarshallingInfo.CharEncoding == CharEncoding.Undefined)
+            {
+                return NoMarshallingInfo.Instance;
+            }
+            else if (_defaultMarshallingInfo.CharEncoding == CharEncoding.Custom)
+            {
+                if (_defaultMarshallingInfo.StringMarshallingCustomType is not null)
+                {
+                    CountInfo countInfo = NoCountInfo.Instance;
+                    if (useSiteAttributes.TryGetUseSiteAttributeInfo(indirectionDepth, out var useSiteInfo))
+                    {
+                        countInfo = useSiteInfo.CountInfo;
+                    }
+                    return CustomMarshallingInfoHelper.CreateNativeMarshallingInfo(
+                        type,
+                        _defaultMarshallingInfo.StringMarshallingCustomType,
+                        _stringMarshallingCustomAttribute,
+                        useSiteAttributes,
+                        marshallingInfoCallback,
+                        true,
```

Comment:
Perhaps we should use a named argument here. Also, where it is used with `false`.